src/cbmc/cbmc_parse_options.cpp

```diff
@@ -470,47 +472,93 @@ int cbmc_parse_optionst::doit()
   register_languages();
 
   if(cmdline.isset("test-preprocessor"))
-    return test_c_preprocessor(ui_message_handler)?8:0;
+    return test_c_preprocessor(get_message_handler())?8:0;
 
   if(cmdline.isset("preprocess"))
   {
     preprocessing();
     return 0; // should contemplate EX_OK from sysexits.h
   }
 
-  goto_functionst goto_functions;
+  if(cmdline.isset("show-parse-tree"))
+  {
+    if(cmdline.args.size()!=1 ||
+       is_goto_binary(cmdline.args[0]))
+    {
+      error() << "Please give exactly one source file" << eom;
+      return 6;
+    }
+
+    std::string filename=cmdline.args[0];
+
+    #ifdef _MSC_VER
+    std::ifstream infile(widen(filename));
+    #else
+    std::ifstream infile(filename);
+    #endif
+
+    if(!infile)
+    {
+      error() << "failed to open input file `"
+              << filename << "'" << eom;
+      return 6;
+    }
+
+    std::unique_ptr<languaget> language=
+      get_language_from_filename(filename);
+
+    if(language==nullptr)
+    {
+      error() << "failed to figure out type of file `"
+              <<  filename << "'" << eom;
+      return 6;
+    }
+
+    language->get_language_options(cmdline);
+    language->set_message_handler(get_message_handler());
+
+    status() << "Parsing " << filename << eom;
+
+    if(language->parse(infile, filename))
+    {
+      error() << "PARSING ERROR" << eom;
+      return 6;
+    }
 
-  expr_listt bmc_constraints;
+    language->show_parse(std::cout);
```

Comment:
Can this entire "show-parse-tree" handling be moved to a separate function so that it can easily be used from other tools accepting source files? At least `symex` come to my mind.

Reply:
Yes, but seems independent job.